File: go/常用算法/146. LRU缓存机制.go
package algorithem

type LRUCache struct {
	size     int
	capacity int
	cache    map[int]*DNode
	head     *DNode
	tail     *DNode
}

type DNode struct {
	key   int
	value int
	pre   *DNode
	next  *DNode
}

func initDNode(key, value int) *DNode {
	return &DNode{
		key:   key,
		value: value,
	}
}

func Constructor(capacity int) LRUCache {
	l := LRUCache{
		size:     0,
		capacity: capacity,
		cache:    make(map[int]*DNode),
		head:     initDNode(0, 0),
		tail:     initDNode(0, 0),
	}
	l.head.next = l.tail
	l.tail.pre = l.head
	return l
}

func (this *LRUCache) Get(key int) int {
	if _, ok := this.cache[key]; !ok {
		return -1
	}
	node := this.cache[key]
	this.moveToHead(node)
	return node.value
}

func (this *LRUCache) Put(key int, value int) {
	if _, ok := this.cache[key]; !ok {
		node := &DNode{key: key, value: value}
		this.cache[key] = node
		this.addToHead(node)
		this.size += 1
		if this.size > this.capacity {
			deleted := this.deleteTail()
			delete(this.cache, deleted.key)
			this.size -= 1
		}
	} else {
		node := this.cache[key]
		node.value = value
		this.moveToHead(node)
	}
}

func (this *LRUCache) deleteNode(node *DNode) {
	node.next.pre = node.pre
	node.pre.next = node.next
}

func (this *LRUCache) deleteTail() *DNode {
	node := this.tail.pre
	this.deleteNode(node)
	return node
}

func (this *LRUCache) moveToHead(node *DNode) {
	this.deleteNode(node)
	this.addToHead(node)
}

func (this *LRUCache) addToHead(node *DNode) {
	node.next = this.head.next
	this.head.next.pre = node
	node.pre = this.head
	this.head.next = node
}

File: python/lru/146. LRU缓存机制.py
class DNode:
    def __init__(self,key=0,value=0):
        self.pre = None
        self.next = None
        self.key = key
        self.value = value

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.cache = dict()
        self.capacity = capacity
        self.size = 0

        self.head = DNode()
        self.tail = DNode()
        self.head.next = self.tail
        self.tail.pre = self.head


    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self.moveToHead(node)
        return node.value


    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key in self.cache:
            node = self.cache[key]
            self.moveToHead(node)
            node.value = value
        else:
            node = DNode(key,value)
            self.addToHead(node)
            self.cache[key] = node
            self.size += 1
            if self.capacity < self.size:
                node  = self.deleteTail()
                self.size -= 1
                self.cache.pop(node.key)

    def moveToHead(self,node):
        self.deleNode(node)
        self.addToHead(node)

    def deleNode(self,node):
        node.pre.next = node.next
        node.next.pre = node.pre

    def addToHead(self,node):
        node.next = self.head.next
        node.next.pre = node
        self.head.next = node
        node.pre = self.head

    def deleteTail(self):
        node = self.tail.pre
        self.deleNode(node)
        return node


cache = LRUCache(2)

cache.put(1, 1)
cache.put(2, 2)
print(cache.get(1))
cache.put(3, 3)
print(cache.get(2))
cache.put(4, 4)
print(cache.get(1))
print(cache.get(3))
print(cache.get(4))

